File: skill/query-gen/java/QueryGenCLI.java
import java.util.*;

/**
 * Query Generator CLI - Java 命令行工具
 */
public class QueryGenCLI {

    public static void main(String[] args) {
        if (args.length == 0 || args[0].equals("--help") || args[0].equals("-h")) {
            printHelp();
            return;
        }

        QueryGeneratorV2 generator = new QueryGeneratorV2();

        // 解析参数
        Map<String, String> params = parseArgs(args);

        // 处理 --list
        if (params.containsKey("list") || params.containsKey("list-personas")) {
            System.out.println("=".repeat(60));
            System.out.println("支持的人设:");
            System.out.println("=".repeat(60));
            for (String persona : generator.getAllPersonaNames()) {
                System.out.println("  - " + persona);
            }
            System.out.println();
        }

        if (params.containsKey("list") || params.containsKey("list-scenes")) {
            System.out.println("=".repeat(60));
            System.out.println("支持的场景:");
            System.out.println("=".repeat(60));
            for (String scene : generator.getAllSceneNames()) {
                System.out.println("  - " + scene);
            }
            System.out.println();
        }

        if (params.containsKey("list") || params.containsKey("list-personas") || params.containsKey("list-scenes")) {
            return;
        }

        // 验证参数
        boolean multiMode = params.containsKey("multi");
        String format = params.getOrDefault("format", "text");
        int count = Integer.parseInt(params.getOrDefault("count", "1"));

        if (!multiMode) {
            // 单人群模式
            String persona = params.get("persona");
            String scene = params.get("scene");

            if (persona == null || scene == null) {
                System.err.println("错误: 非多人群模式下，--persona 和 --scene 是必需的");
                System.exit(1);
            }

            List<QueryGeneratorV2.GeneratedQuery> queries = generator.generateBatch(persona, scene, count);
            String output = formatBatchOutput(queries, format);

            if (params.containsKey("output")) {
                System.out.println(output);
                System.out.println("结果已保存到: " + params.get("output"));
            } else {
                System.out.println(output);
            }

        } else {
            // 多人群模式
            String personasStr = params.get("personas");
            String scene = params.get("scene");

            if (personasStr == null || scene == null) {
                System.err.println("错误: 多人群模式下，--personas 和 --scene 是必需的");
                System.exit(1);
            }

            List<String> personaList = Arrays.asList(personasStr.split(","));
            Map<String, List<QueryGeneratorV2.GeneratedQuery>> results =
                generator.generateMultiPersona(personaList, scene, count);
            String output = formatMultiOutput(results, format);

            if (params.containsKey("output")) {
                System.out.println(output);
                System.out.println("结果已保存到: " + params.get("output"));
            } else {
                System.out.println(output);
            }
        }
    }

    private static Map<String, String> parseArgs(String[] args) {
        Map<String, String> params = new HashMap<>();

        for (int i = 0; i < args.length; i++) {
            String arg = args[i];
            switch (arg) {
                case "--persona":
                case "-p":
                    params.put("persona", args[++i]);
                    break;
                case "--scene":
                case "-s":
                    params.put("scene", args[++i]);
                    break;
                case "--count":
                case "-c":
                    params.put("count", args[++i]);
                    break;
                case "--multi":
                case "-m":
                    params.put("multi", "true");
                    break;
                case "--personas":
                    params.put("personas", args[++i]);
                    break;
                case "--format":
                case "-f":
                    params.put("format", args[++i]);
                    break;
                case "--output":
                case "-o":
                    params.put("output", args[++i]);
                    break;
                case "--list":
                case "-l":
                    params.put("list", "true");
                    break;
                case "--list-personas":
                    params.put("list-personas", "true");
                    break;
                case "--list-scenes":
                    params.put("list-scenes", "true");
                    break;
            }
        }

        return params;
    }

    private static String formatBatchOutput(List<QueryGeneratorV2.GeneratedQuery> queries, String format) {
        if ("json".equals(format)) {
            StringBuilder sb = new StringBuilder();
            sb.append("[\n");
            for (int i = 0; i < queries.size(); i++) {
                QueryGeneratorV2.GeneratedQuery q = queries.get(i);
                sb.append("  {\n");
                sb.append("    \"query\": \"").append(escapeJson(q.getQuery())).append("\",\n");
                sb.append("    \"persona\": \"").append(q.getPersona()).append("\",\n");
                sb.append("    \"scene\": \"").append(q.getScene()).append("\",\n");
                sb.append("    \"intent\": \"").append(q.getIntent()).append("\",\n");
                sb.append("    \"complexity\": \"").append(q.getComplexity()).append("\",\n");
                sb.append("    \"emotionLevel\": \"").append(q.getEmotionLevel()).append("\"\n");
                sb.append("  }");
                if (i < queries.size() - 1) sb.append(",");
                sb.append("\n");
            }
            sb.append("]");
            return sb.toString();

        } else if ("csv".equals(format)) {
            StringBuilder sb = new StringBuilder();
            sb.append("query,persona,scene,intent,complexity,emotionLevel\n");
            for (QueryGeneratorV2.GeneratedQuery q : queries) {
                sb.append("\"").append(q.getQuery()).append("\",");
                sb.append(q.getPersona()).append(",");
                sb.append(q.getScene()).append(",");
                sb.append(q.getIntent()).append(",");
                sb.append(q.getComplexity()).append(",");
                sb.append(q.getEmotionLevel()).append("\n");
            }
            return sb.toString();

        } else { // text
            StringBuilder sb = new StringBuilder();
            sb.append("=".repeat(60)).append("\n");
            sb.append("生成结果: ").append(queries.get(0).getPersona())
              .append(" - ").append(queries.get(0).getScene()).append("\n");
            sb.append("=".repeat(60)).append("\n");

            for (int i = 0; i < queries.size(); i++) {
                QueryGeneratorV2.GeneratedQuery q = queries.get(i);
                sb.append("\n【").append(i + 1).append("】\n");
                sb.append("  Query: ").append(q.getQuery()).append("\n");
                sb.append("  意图: ").append(q.getIntent()).append("\n");
                sb.append("  复杂度: ").append(q.getComplexity()).append("\n");
                sb.append("  情感强度: ").append(q.getEmotionLevel()).append("\n");
                sb.append("  评测重点: ").append(String.join(", ", q.getExpectedFocus())).append("\n");
            }

            sb.append("\n").append("=".repeat(60)).append("\n");
            return sb.toString();
        }
    }

    private static String formatMultiOutput(Map<String, List<QueryGeneratorV2.GeneratedQuery>> results, String format) {
        if ("json".equals(format)) {
            StringBuilder sb = new StringBuilder();
            sb.append("{\n");
            int personaCount = 0;
            for (Map.Entry<String, List<QueryGeneratorV2.GeneratedQuery>> entry : results.entrySet()) {
                sb.append("  \"").append(entry.getKey()).append("\": [\n");
                List<QueryGeneratorV2.GeneratedQuery> queries = entry.getValue();
                for (int i = 0; i < queries.size(); i++) {
                    QueryGeneratorV2.GeneratedQuery q = queries.get(i);
                    sb.append("    {\"query\": \"").append(escapeJson(q.getQuery())).append("\"}");
                    if (i < queries.size() - 1) sb.append(",");
                    sb.append("\n");
                }
                sb.append("  ]");
                if (++personaCount < results.size()) sb.append(",");
                sb.append("\n");
            }
            sb.append("}");
            return sb.toString();

        } else { // text
            StringBuilder sb = new StringBuilder();
            sb.append("=".repeat(60)).append("\n");
            sb.append("多人群对比结果").append("\n");
            sb.append("=".repeat(60)).append("\n");

            for (Map.Entry<String, List<QueryGeneratorV2.GeneratedQuery>> entry : results.entrySet()) {
                sb.append("\n【").append(entry.getKey()).append("】\n");
                List<QueryGeneratorV2.GeneratedQuery> queries = entry.getValue();
                for (int i = 0; i < queries.size(); i++) {
                    sb.append("  ").append(i + 1).append(". ")
                      .append(queries.get(i).getQuery()).append("\n");
                }
            }

            sb.append("\n").append("=".repeat(60)).append("\n");
            return sb.toString();
        }
    }

    private static String escapeJson(String s) {
        return s.replace("\\", "\\\\")
                .replace("\"", "\\\"")
                .replace("\n", "\\n")
                .replace("\r", "\\r")
                .replace("\t", "\\t");
    }

    private static void printHelp() {
        System.out.println("Query Generator CLI - Java 命令行工具");
        System.out.println();
        System.out.println("用法: java QueryGenCLI [选项]");
        System.out.println();
        System.out.println("选项:");
        System.out.println("  --persona, -p <名称>     人设名称");
        System.out.println("  --scene, -s <名称>       场景名称");
        System.out.println("  --count, -c <数量>       生成数量 (默认: 1)");
        System.out.println("  --multi, -m              多人群对比模式");
        System.out.println("  --personas <列表>        多个人设，用逗号分隔");
        System.out.println("  --format, -f <格式>      输出格式: text/json/csv (默认: text)");
        System.out.println("  --output, -o <路径>      输出文件路径");
        System.out.println("  --list, -l               列出所有人设和场景");
        System.out.println("  --list-personas          列出所有人设");
        System.out.println("  --list-scenes            列出所有场景");
        System.out.println("  --help, -h               显示帮助");
        System.out.println();
        System.out.println("示例:");
        System.out.println("  # 列出所有人设和场景");
        System.out.println("  java QueryGenCLI --list");
        System.out.println();
        System.out.println("  # 生成单个 Query");
        System.out.println("  java QueryGenCLI --persona \"新手焦虑型妈妈\" --scene \"医疗健康\"");
        System.out.println();
        System.out.println("  # 批量生成 10 条");
        System.out.println("  java QueryGenCLI -p \"新手焦虑型妈妈\" -s \"医疗健康\" -c 10");
        System.out.println();
        System.out.println("  # 多人群对比");
        System.out.println("  java QueryGenCLI --multi --personas \"新手焦虑型妈妈,谨慎保守型老人\" -s \"医疗健康\" -c 5");
        System.out.println();
        System.out.println("  # 输出 JSON 格式");
        System.out.println("  java QueryGenCLI -p \"新手焦虑型妈妈\" -s \"医疗健康\" -c 5 -f json");
    }
}

File: skill/query-gen/python/query_gen_cli.py
#!/usr/bin/env python3
# -*- coding: utf-8 -*-
"""
Query Generator CLI - 支持命令行参数调用的 Skill 示例
"""

import argparse
import json
import sys
from pathlib import Path

# 将项目路径添加到 sys.path
sys.path.insert(0, str(Path(__file__).parent))

from query_generator import QueryGenerator


def main():
    parser = argparse.ArgumentParser(
        description="基于数字人设自动生成搜索 Query",
        formatter_class=argparse.RawDescriptionHelpFormatter,
        epilog="""
示例:
  # 生成单个 Query
  python query_gen_cli.py --persona "新手焦虑型妈妈" --scene "医疗健康"

  # 批量生成 10 条
  python query_gen_cli.py --persona "新手焦虑型妈妈" --scene "医疗健康" --count 10

  # 多人群对比
  python query_gen_cli.py --multi --personas "新手焦虑型妈妈,谨慎保守型老人" --scene "医疗健康" --count 5

  # 列出所有人设和场景
  python query_gen_cli.py --list

  # 输出 JSON 格式
  python query_gen_cli.py --persona "新手焦虑型妈妈" --scene "医疗健康" --count 5 --format json
        """
    )

    # 主要参数
    parser.add_argument("--persona", "-p", type=str, help="人设名称")
    parser.add_argument("--scene", "-s", type=str, help="场景名称")
    parser.add_argument("--count", "-c", type=int, default=1, help="生成数量 (默认: 1)")
    parser.add_argument("--urgency", "-u", type=str, help="紧急程度 (可选)")

    # 多人群对比模式
    parser.add_argument("--multi", "-m", action="store_true", help="多人群对比模式")
    parser.add_argument("--personas", type=str, help="多个人设，用逗号分隔")

    # 输出格式
    parser.add_argument("--format", "-f", choices=["text", "json", "csv"], default="text",
                        help="输出格式 (默认: text)")
    parser.add_argument("--output", "-o", type=str, help="输出文件路径 (可选)")

    # 信息查询
    parser.add_argument("--list", "-l", action="store_true", help="列出所有人设和场景")
    parser.add_argument("--list-personas", action="store_true", help="列出所有人设")
    parser.add_argument("--list-scenes", action="store_true", help="列出所有场景")

    args = parser.parse_args()

    # 初始化生成器
    generator = QueryGenerator()

    # 处理信息查询
    if args.list or args.list_personas:
        print("=" * 60)
        print("支持的人设:")
        print("=" * 60)
        for persona in generator.get_all_persona_names():
            print(f"  - {persona}")
        print()

    if args.list or args.list_scenes:
        print("=" * 60)
        print("支持的场景:")
        print("=" * 60)
        for scene in generator.get_all_scene_names():
            print(f"  - {scene}")
        print()

    if args.list or args.list_personas or args.list_scenes:
        return

    # 验证参数
    if not args.multi and (not args.persona or not args.scene):
        parser.error("非多人群模式下，--persona 和 --scene 是必需的")

    if args.multi and (not args.personas or not args.scene):
        parser.error("多人群模式下，--personas 和 --scene 是必需的")

    # 生成 Query
    if args.multi:
        # 多人群对比模式
        persona_list = [p.strip() for p in args.personas.split(",")]
        results = generator.generate_multi_persona(
            persona_list, args.scene, args.count
        )
        output = format_multi_output(results, args.format)
    else:
        # 单人群批量生成
        queries = generator.generate_batch(args.persona, args.scene, args.count)
        output = format_batch_output(queries, args.format)

    # 输出结果
    if args.output:
        with open(args.output, "w", encoding="utf-8") as f:
            f.write(output)
        print(f"结果已保存到: {args.output}")
    else:
        print(output)


def format_batch_output(queries, format_type):
    """格式化批量输出"""
    if format_type == "json":
        data = [q.to_dict() for q in queries]
        return json.dumps(data, ensure_ascii=False, indent=2)

    elif format_type == "csv":
        lines = ["query,persona,scene,intent,complexity,emotion_level,expected_focus"]
        for q in queries:
            focus = "|".join(q.expected_focus)
            lines.append(f'"{q.query}",{q.persona},{q.scene},{q.intent},{q.complexity},{q.emotion_level},"{focus}"')
        return "\n".join(lines)

    else:  # text
        lines = ["=" * 60]
        lines.append(f"生成结果: {queries[0].persona} - {queries[0].scene}")
        lines.append("=" * 60)

        for i, q in enumerate(queries, 1):
            lines.append(f"\n【{i}】")
            lines.append(f"  Query: {q.query}")
            lines.append(f"  意图: {q.intent}")
            lines.append(f"  复杂度: {q.complexity}")
            lines.append(f"  情感强度: {q.emotion_level}")
            lines.append(f"  评测重点: {', '.join(q.expected_focus)}")

        lines.append("\n" + "=" * 60)
        return "\n".join(lines)


def format_multi_output(results, format_type):
    """格式化多人群输出"""
    if format_type == "json":
        data = {
            persona: [q.to_dict() for q in queries]
            for persona, queries in results.items()
        }
        return json.dumps(data, ensure_ascii=False, indent=2)

    elif format_type == "csv":
        lines = ["query,persona,scene,intent,complexity,emotion_level,expected_focus"]
        for persona, queries in results.items():
            for q in queries:
                focus = "|".join(q.expected_focus)
                lines.append(f'"{q.query}",{q.persona},{q.scene},{q.intent},{q.complexity},{q.emotion_level},"{focus}"')
        return "\n".join(lines)

    else:  # text
        lines = ["=" * 60]
        lines.append("多人群对比结果")
        lines.append("=" * 60)

        for persona, queries in results.items():
            lines.append(f"\n【{persona}】")
            for i, q in enumerate(queries, 1):
                lines.append(f"  {i}. {q.query}")

        lines.append("\n" + "=" * 60)
        return "\n".join(lines)


if __name__ == "__main__":
    main()
